src/sage/geometry/cone_critical_angles.py

```diff
@@ -0,0 +1,998 @@
+    therefore crash it by feeding it invalid input like an
+    inadmissible cone::
+
+        sage: from sage.geometry.cone_critical_angles import max_angle
+        sage: K = cones.trivial(3)
+        sage: max_angle(K,K,True,0)
+        Traceback (most recent call last):
+        ...
+        IndexError: list index out of range
+    """
+    # The lattice dimensions of P and Q are guaranteed to be equal
+    # because the cone method checks it before calling us.
+    n = P.lattice_dim()
+
+    ring = RDF
+    if exact:
+        ring = AA
+        # For some reason we can go RR -> QQ -> AA, but not
+        # straight from RR to AA.
+        epsilon = QQ(epsilon)
+    epsilon = ring(epsilon)
+
+    # First check if P is contained in the dual of Q. Keep track of
+    # the minimum inner product (and associated vectors) while doing
+    # so; then if P is contained in dual(Q), we just return the pair
+    # with the smallest inner product.
+    gs = [g.change_ring(ring).normalized() for g in P]
+    Q_is_P = (P == Q) # This is used again later
+    if Q_is_P:
+        hs = gs
+    else:
+        hs = [h.change_ring(ring).normalized() for h in Q]
+
+    (M, min_ip, min_u, min_v) = compute_gevp_M(gs,hs)
+
+    if min_ip >= 0: # The maximal angle is acute!
+        return (arccos(min_ip), min_u, min_v)
+
+    # Also check to see if the maximal angle is pi. In particular this
+    # is true when either P or Q is the entire ambient space.
+    P_and_negative_Q = P.intersection(-Q)
+    if not (P_and_negative_Q.is_trivial()):
+        u = P_and_negative_Q.ray(0).change_ring(ring).normalized()
+        v = -u
+        return (pi, u, v)
+
+    # When P == Q, GG and HH are both just M. We rule out the
+    # cardinality ``n`` in the index sets because it will eventually
+    # result in a GEVP whose only solutions are lambda in {-1,0,1};
+    # none of which we want! We rule out 0 and 1 with the acute check,
+    # and -1 with the pi (P_and_negative_Q) check.
+    #
+    # It's VERY IMPORTANT that we construct lists from the index set
+    # generators, because we're going to use them in a nested loop!
+    G = matrix.column(gs)
+    G_index_sets = [s for s in gevp_licis(G) if not len(s) == n]
+
+    if Q_is_P:
+        GG = M
+        H = G
+        HH = M
+        H_index_sets = G_index_sets
+    else:
+        GG = G.transpose()*G
+        H = matrix.column(hs)
+        HH = H.transpose()*H
+        H_index_sets = [s for s in gevp_licis(H) if not len(s) == n]
+
+    # Keep track of the (cos-theta, xi, eta, multiplicity) tuples with
+    # multiplicity > 1. These are only a problem if they could
+    # potentially be maximal. Therefore, we want to inspect them AFTER
+    # we've checked all of the multiplicity=1 angles and found the
+    # largest. This allows us to ignore most of the problematic
+    # eigenspaces, independent of the order in which we run through I,J.
+    big_eigenspaces = []
+
+    for I in G_index_sets:
+        G_I = G[range(n),I]
```

Comment:
Wouldn't `G.matrix_from_columns(I)` be faster here? Similar for other places below.

Reply:
Yep. I didn't know that method was there.